bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/firmware/FirmwareRegistryImpl.java

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.thing.internal.firmware;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.stream.Collectors;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.i18n.LocaleProvider;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.binding.firmware.Firmware;
+import org.eclipse.smarthome.core.thing.firmware.FirmwareProvider;
+import org.eclipse.smarthome.core.thing.firmware.FirmwareRegistry;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Default implementation of {@link FirmwareRegistry}.
+ *
+ * @author Thomas Höfer - Initial contribution
+ * @author Dimitar Ivanov - The firmwares are provided by thing and version
+ */
+@Component(immediate = true, service = FirmwareRegistry.class)
+@NonNullByDefault
+public final class FirmwareRegistryImpl implements FirmwareRegistry {
+
+    private final Logger logger = LoggerFactory.getLogger(FirmwareRegistryImpl.class);
+
+    private final List<FirmwareProvider> firmwareProviders = new CopyOnWriteArrayList<>();
+
+    private @Nullable LocaleProvider localeProvider;
```

Comment:
As the `LocaleProvider` is a static mandatory reference, you could annotate it as `@NonNullByDefault({})` (instead of `@Nullable`), as indicated here: https://www.eclipse.org/smarthome/documentation/development/conventions.html#null-annotations

Reply:
You are absolutely right! I will also add `@NonNullByDefault({})` into the FirmwareUpdateServiceImpl, where I have several static references as well.